### PR TITLE
[AMD][AgentX] DeepSeek V4 multi-node Agentic framework and recipe / DeepSeek V4 多节点 Agentic 框架与配方

### PR DESCRIPTION
**Pull request by ichbinblau**
## Summary

Add end-to-end agentic benchmarking for DeepSeek-V4-Pro FP4 on MI355X using multi-node disaggregated SGLang and HiCache host-DRAM KV offloading.

This PR now focuses on the recipe and runtime implementation:

- add the DSV4 MI355X disaggregated Agentic benchmark entrypoint;
- add the AIPerf trace-replay path and Slurm/container wiring;
- configure SGLang disaggregated serving and HiCache host-pool settings;
- register the model, benchmark configuration, runner launch behavior, and changelog trigger;
- expose SGLang host-cache hit rate in the existing structured Agentic result aggregation;
- retain only local patch code that does not have a merged upstream equivalent.

## Dependency

Depends on #2141, which separately enables `kv-offloading` and `kv-offload-backend` propagation for multi-node Agentic sweep matrices. The shared `utils/matrix_logic` changes are intentionally not included in this PR.

Until #2141 is merged, the configuration in this PR declares `kv-offloading: dram` and `kv-offload-backend: hicache`, but current `main` still forces multi-node Agentic matrix entries to `kv-offloading: none`.

## Patch policy

Patch and overlay code corresponding to merged upstream SGLang changes has been removed. The remaining patch artifacts are limited to code without a merged upstream equivalent, including the disaggregated prefill bootstrap desynchronization fix and the reference-only decode TP queue agreement patch.

## Validation

- Shell and Slurm scripts pass `bash -n`.
- `configs/amd-master.yaml` and `perf-changelog.yaml` parse successfully.
- `PYTHONPATH=. pytest -q utils/agentic/aggregation/test_process_agentic_result.py` — 26 passed.
- Changed Python aggregation files compile successfully.
- `git diff --check` passes.

## 中文说明

本 PR 为 MI355X 上的 DeepSeek-V4-Pro FP4 增加端到端 Agentic 基准测试能力，采用多节点 SGLang 分离式推理，并通过 HiCache 使用主机 DRAM 进行 KV 缓存卸载。

本 PR 当前仅聚焦配方和运行时实现：

- 新增 DSV4 MI355X 分离式 Agentic 基准测试入口；
- 增加 AIPerf trace replay 路径以及 Slurm/容器运行链路；
- 配置 SGLang 分离式服务和 HiCache 主机缓存池参数；
- 注册模型、基准测试配置、runner 启动行为和 changelog 触发项；
- 在现有 Agentic 结构化结果聚合中暴露 SGLang 主机缓存命中率；
- 仅保留尚无已合并上游实现的本地补丁代码。

## 依赖关系

本 PR 依赖 #2141。#2141 单独为多节点 Agentic 扫描矩阵增加 `kv-offloading` 和 `kv-offload-backend` 的透传能力。本 PR 已移除所有共享的 `utils/matrix_logic` 改动。

在 #2141 合并之前，本 PR 的配置虽然声明了 `kv-offloading: dram` 和 `kv-offload-backend: hicache`，但当前 `main` 仍会把多节点 Agentic 矩阵项强制转换为 `kv-offloading: none`。

## 补丁策略

与已合并 SGLang 上游改动对应的补丁和 overlay 代码均已移除。当前仅保留尚无已合并上游等价实现的本地补丁，包括分离式 prefill bootstrap 不同步修复，以及仅作参考、默认不启用的 decode TP queue agreement 补丁。

## 验证

Shell 和 Slurm 脚本均通过 `bash -n`；`configs/amd-master.yaml` 与 `perf-changelog.yaml` 可正常解析；Agentic 聚合测试 26 项全部通过；相关 Python 文件编译检查和 `git diff --check` 均通过。

### CODE REVIEW

File: configs/amd-master.yaml [hunk 1]
```diff
@@ -3041,3 +3041,54 @@ minimaxm3-fp8-mi325x-vllm-agentic:
       - { tp: 4, ep: 4, kv-offloading: dram, kv-offload-backend: native, conc-list: [3, 4, 5, 6, 7, 8, 9, 10, 12, 14, 16] }
       - { tp: 8, ep: 8, kv-offloading: dram, kv-offload-backend: native, conc-list: [10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 32] }
       - { tp: 8, ep: 8, dp-attn: true, kv-offloading: dram, kv-offload-backend: native, conc-list: [24, 32, 36, 40, 44, 48, 52, 56, 60, 64, 72, 80, 96] }
+
+dsv4-fp4-mi355x-sglang-disagg-agentic-hicache:
+  image: rocm/sgl-dev:v0.5.14-rocm720-mi35x-20260706
```
Comment (functionstackx) [inline configs/amd-master.yaml]: do u have upstream image @ichbinblau

**Comment by Duyi-Wang** [comment]
The patches in this PR are derived from the following upstream SGLang PRs.
https://github.com/sgl-project/sglang/pull/30374
https://github.com/sgl-project/sglang/pull/30333
https://github.com/sgl-project/sglang/pull/30339
https://github.com/sgl-project/sglang/pull/29417

File: configs/amd-master.yaml (hunk 1 above)
Comment (ichbinblau) [reply configs/amd-master.yaml]: We have bumped the image to `lmsysorg/sglang-rocm:v0.5.14-rocm720-mi35x-20260706`

File: benchmarks/multi_node/amd_utils/patches/decode_tp_queue_agree.patch [hunk 2]
```diff
@@ -0,0 +1,107 @@
+--- a/python/sglang/srt/disaggregation/decode.py	2026-06-25 02:44:11.274035348 +0000
++++ b/python/sglang/srt/disaggregation/decode.py	2026-06-25 02:44:11.276035310 +0000
+@@ -1616,8 +1616,57 @@
+         )
+         kv_manager._staging_handler = self.staging_handler
+
++    def _agree_and_order_queue(self) -> Tuple[List["DecodeRequest"], List["DecodeRequest"]]:
```
Comment (cquil11) [inline benchmarks/multi_node/amd_utils/patches/decode_tp_queue_agree.patch]: We will not accept any patches. I have removed all patches that have since been merged into upstream sglang main. Please update nightly image tomorrow. 

However, this patch remains. Please remove it / open upstream pr.

**Review by cquil11 — CHANGES_REQUESTED** [review]

**Comment by cquil11** [comment]
@ichbinblau Thank you for this awesome work. Please see the changes I requested. Also, please re run full-sweep-enabled after making these changes. 

+viz @chunfangamd

File: configs/amd-master.yaml (hunk 1 above)
Comment (billishyahao) [reply configs/amd-master.yaml]: Yes we will switch to lmsysorg/sglang-rocm:v0.5.14-rocm720-mi35x-20260710 where included all 4 patches

File: benchmarks/multi_node/amd_utils/patches/decode_tp_queue_agree.patch (hunk 2 above)
Comment (ichbinblau) [reply benchmarks/multi_node/amd_utils/patches/decode_tp_queue_agree.patch]: > We will not accept any patches. I have removed all patches that have since been merged into upstream sglang main. Please update nightly image tomorrow.
> 
> However, this patch remains. Please remove it / open upstream pr.

This patch file has not been used as its caller func has been commented out. I have remove it.

**Review by cquil11 — APPROVED** [review]
LGTM. Definitely needs some cleanup which we will do in a followup. Thanks for the great best effort on first submission. 

https://github.com/SemiAnalysisAI/InferenceX/issues/21

File: benchmarks/multi_node/amd_utils/setup_deps.sh [hunk 3]
```diff
@@ -80,86 +79,46 @@ install_amd_quark() {
...
-# Affects: GLM-5 (NSA attention) and any future model that uses
-# deepgemm_fp8_paged_mqa_logits with Preshuffle=False.
+# resolve_waiting_queue_bootstrap() runs poll_and_all_reduce_attn_cp_tp_group()
+# over `candidates`. The upstream candidate set (all non-aborted waiting reqs)
+# can differ across TP ranks, so some ranks enter the all_reduce while others
+# skip it -> hang. Narrow candidates to optimistic (pending_bootstrap) requests,
+# which is consistent across ranks and is the only set finalize_bootstrap acts on.
 # ---------------------------------------------------------------------------
-patch_gluon_pa_mqa_logits_instr_shape() {
+patch_disagg_prefill_bootstrap_desync() {
```
Comment (cquil11) [inline benchmarks/multi_node/amd_utils/setup_deps.sh]: I'm going to allow this for now as we are early days of agentX and sglang and vllm also have some scheduling patches, but I want to open an issue for you guys to track this and get an upstream fix in sglang. @billishyahao @ichbinblau 
cc @ichbinblau

Comment (cquil11) [reply benchmarks/multi_node/amd_utils/setup_deps.sh]: https://github.com/SemiAnalysisAI/InferenceX/issues/2151

**Review by chunfangamd — APPROVED** [review]
As a PR reviewer and CODEOWNER, I have reviewed this and have:
- [x] Verified that as of the moment of typing this, this is the latest version of [PR_REVIEW_CHECKLIST.md](https://github.com/SemiAnalysisAI/InferenceX/blob/main/docs/PR_REVIEW_CHECKLIST.md)
- [x] Verified that the general code quality meets the InferenceX standard and does not make the code quality any worse. Approve for day 0
- [x] Verified that this PR has passed PR validation. Please link to GitHub Action workflow that shows this. https://github.com/SemiAnalysisAI/InferenceX/actions/runs/29089300938
- [x] Verified that this PR passes evals.  Please link to GitHub Action workflow that shows this. https://github.com/SemiAnalysisAI/InferenceX/actions/runs/29089300938
- [x] Verified that speculative decoding PRs uses chat templates to align the AL distribution to real world
- [x] Verified that the model architecture isn't changed with benchmark hacks like using --hf-overrides to skipping indexer for every x layers on models that don't natively support this. As a general rule, we won't accept optimizations that reduces the number of model architecture FLOPs. Anything that makes that same computation run faster is fair game; FLOPs at lower precisions is fine, given that the config passes private evals. As an general north star princple, we should only use optimizations which is used in production by customers that care about accuracy
- [x] If an company claims that they support vLLM/SGLang as first class LLM inference engines on their hardware, I have verified that the respective vLLM submission made using upstream https://hub.docker.com/u/vllm docker repo, upstream SGLang https://hub.docker.com/u/lmsysorg docker repo. The only exceptions are for new hardware, such as MI455X UALoE72, Vera Rubin NVL72, Rubin NVL8, etc., and for new model architectures where there is an actual reason why vLLM/SGLang does not fundamentally support them yet as supported by vLLM/SGLang community maintainers
- [x] If an company claims that they support vLLM/SGLang as first class upstream in-tree LLM inference engines on their hardware, I have have verified that the respective vLLM/SGLang submission has been made before additional frameworks (TRT-LLM, ATOM, etc.). The only exceptions are for new hardware, such as MI455X UALoE72, Vera Rubin NVL72, Rubin NVL8, etc., and for new model architectures where there is an actual reason why vLLM/SGLang does not fundamentally support them yet.
- [x] Verified that the single-node recipes are similar to the official [vLLM recipes](https://recipes.vllm.ai/) and/or the[SGLang cookbook](https://docs.sglang.io/cookbook/intro):
  - If they are not, I have verified that a PR has been opened in [vLLM recipe repo](https://github.com/vllm-project/recipes) or [SGLang repo](https://github.com/sgl-project/sglang/tree/main/docs_new) and linked it below in the additional detail section:
- [ ] If any of the above criteria cannot reasonably be satisfied, I have provided additional reasoning below.

### Additional detail section:
- insert any additional info here
We will need a SGLang cookbook for the recipe. I approve it for the day 0 support for the AgentX

Signed: @chunfangamd

**Comment by Klaud-Cold** [comment]
## ❌❌❌ **REJECTED** ❌❌❌
@chunfangamd — blocking: no green, *executed* sweep/eval exists on any commit currently in this PR, so the reuse gate (`validate_reusable_run`) will reject it at merge. Everything else is to standard.

❌ Check 1 (sweep+evals on in-PR commit): FAIL — the PR's only commit `f409c366` has `single-node */`, `multi-node */`, and `eval /` all `skipped`. The signed-off run [29089300938](https://github.com/SemiAnalysisAI/InferenceX/actions/runs/29089300938) executed on `bb95016b`, which is not in `GET /pulls/2127/commits`. Fix: run (or re-anchor via `/reuse-sweep-run`) a passing full sweep on a commit currently in this PR.
❌ Check 2 (evals accuracy): FAIL — no eval run exists on an in-PR commit to verify accuracy against; re-derivable only after Check 1 is satisfied.
✅ Check 0 (CODEOWNER): PASS — `@chunfangamd` owns `configs/amd-master.yaml`; all other paths fall to the `* @InferenceX/core` catch-all, covered.
➖ Check 3 (recipe link): N/A — disaggregated/multi-node submission (`sglang-disagg`, `multinode: true`, `disagg: true`, files under `benchmarks/multi_node/**`); the recipe-link requirement applies to single-node recipes only.
✅ Check 4 (reuse-sweep command): PASS — `/reuse-sweep-run` posted by `cquil11` (MEMBER).
✅ Check 5 (latest checklist template): PASS — all current template items present and checked; only the conditional "if criteria cannot be satisfied" escape-hatch item left unchecked, which is appropriate.
✅ Check 6 (upstream image + engine order): PASS — `lmsysorg/sglang-rocm:v0.5.14-rocm720-mi35x-20260710` is upstream SGLang; SGLang framework, so engine-first ordering not triggered.
✅ Check 7 (no arch/FLOP hacks): PASS — only kernel-routing/cache and MoRI/HiCache tuning knobs; no `--hf-overrides`/`--json-model-override-args` reducing architecture FLOPs.
➖ Check 8 (spec-decode chat templates): N/A — `spec-decoding: none`, `DECODE_MTP_SIZE=0`; no speculative-decoding changes.

File: benchmarks/multi_node/amd_utils/setup_deps.sh (hunk 3 above)
Comment (ichbinblau) [reply benchmarks/multi_node/amd_utils/setup_deps.sh]: > #2151

Thanks, the patch is not necessary in the new image, too. Will remove it.
